azurelinuxagent/ga/update.py

```diff
@@ -540,30 +551,30 @@ def handle_updates_for_requested_version():
 
         self.__upgrade_agent_if_permitted()
 
-    def __goal_state_updated(self, incarnation):
+    def _processing_new_incarnation(self):
+        """
+        True if we are currently processing a new incarnation (i.e. WireServer goal state)
+        """
+        return self._goal_state is not None and self._goal_state.incarnation != self._last_incarnation
+
+    def _processing_new_extensions_goal_state(self):
         """
-        This function returns if the Goal State updated.
-        We currently rely on the incarnation number to determine that; i.e. if it changed from the last processed GS
+        True if we are currently processing a new extensions goal state
         """
-        # TODO: This check should be based on the ExtensionsGoalState.id property
-        #  (this property abstracts incarnation/etag logic based on the delivery pipeline of the Goal State)
-        return incarnation != self.last_incarnation
+        return self._goal_state is not None and self._goal_state.extensions_goal_state.id != self._last_extensions_gs_id
 
     def _process_goal_state(self, exthandlers_handler, remote_access_handler):
+        try:
+            protocol = exthandlers_handler.protocol
 
-        protocol = exthandlers_handler.protocol
-        if not self._try_update_goal_state(protocol):
-            self._heartbeat_update_goal_state_error_count += 1
-            # We should have a cached goal state here, go ahead and report status for that.
-            self._report_status(exthandlers_handler, incarnation_changed=False)
-            return
+            # update self._goal_state
+            self._try_update_goal_state(protocol)
 
-        # Update the Guest Agent if a new version is available
-        self.__update_guest_agent(protocol)
-        incarnation = protocol.get_incarnation()
+            # Update the Guest Agent if a new version is available
+            if self._goal_state is not None:
+                self.__update_guest_agent(protocol)
```

Comment:
Not having `return` on try_update_goal_state failure case, changes the existing behavior of update guest agent.

Example previous goal state received and updated `self._goal_state is not None` which processed update guest agent. All good at this point. On a current goal state, we try to do `try_update_goal_state ` which is failed and self._goal_state is not updated and will have previous goal sate. So now the `update guest agent `processed on previous goal state information which we already done as part of previous goal state. Do we need to run update guest agent without incarnation changes?

Reply:
yes, we need to agent update and status report even if no incarnation change; the previous code was skipping agent update so this change was intentional